lib/smack/Prelude.cpp

```diff
@@ -280,39 +290,88 @@ std::string getBytewisePointerStorageStr() {
   return s.str();
 }
 
-std::string Prelude::getPrelude() {
-  const std::vector<unsigned> INTEGER_SIZES {
-    1, 5, 6, 8, 16, 24, 32, 40, 48, 56, 64, 80, 88, 96, 128, 160, 256
-  };
+const std::vector<unsigned> INTEGER_SIZES {
+  1, 5, 6, 8, 16, 24, 32, 40, 48, 56, 64, 80, 88, 96, 128, 160, 256
+};
 
-  const std::vector<unsigned> REF_CONSTANTS {
-    0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 1024
-  };
+const std::vector<unsigned> REF_CONSTANTS {
+  0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 1024
+};
 
-  const std::vector<std::string> BINARY_OPS {
+void Prelude::generateBinaryOps(std::stringstream& s, const unsigned size) {
+  static const std::vector<std::string> BINARY_OPS {
     "add", "sub", "mul", "sdiv", "smod", "srem", "udiv",
-    "urem", "shl", "lshr", "ashr", "and", "or", "xor", "nand"
+      "urem", "shl", "lshr", "ashr", "and", "or", "xor", "nand"
   };
+  static const std::vector<BinExpr::Binary> BINARY_OPS_BODY {
+    BinExpr::Plus, BinExpr::Minus, BinExpr::Times
+  };
+  for (size_t i = 0; i < BINARY_OPS.size(); ++i) {
+    const std::string &op = BINARY_OPS[i];
+    s << bvbuiltinBinaryOp(size, "bv" + op, "$" + op) << "\n";
+    if (i <= 2) {
+      s << inlineBinaryOp(size, "$" + op,
+          new BinExpr(BINARY_OPS_BODY[i], makeVarExpr(1), makeVarExpr(2)))
+        << "\n";
+    } else if (i <= 7) {
+      s << builtinBinaryOp(size, op.substr(1), "$" + op) << "\n";
+    } else {
+      s << uninterpretedBinaryOp(size, "$" + op) << "\n";
+    }
+  }
+}
 
-  const std::vector<std::string> COMP_OPS {
+void Prelude::generateRelationalOps(std::stringstream& s, const unsigned size) {
+  static const std::vector<std::string> COMP_OPS {
     "ule", "ult", "uge", "ugt", "sle", "slt", "sge", "sgt"
   };
 
-  const std::vector<std::string> EQ_OPS {"eq", "ne"};
+  static const std::vector<std::string> EQ_OPS {"eq", "ne"};
 
-  const std::vector<BinExpr::Binary> BINARY_OPS_BODY {
-    BinExpr::Plus, BinExpr::Minus, BinExpr::Times
-  };
 
-  const std::vector<BinExpr::Binary> COMP_OPS_BODY {
+  static const std::vector<BinExpr::Binary> COMP_OPS_BODY {
     BinExpr::Lte, BinExpr::Lt, BinExpr::Gte, BinExpr::Gt,
-    BinExpr::Lte, BinExpr::Lt, BinExpr::Gte, BinExpr::Gt
+      BinExpr::Lte, BinExpr::Lt, BinExpr::Gte, BinExpr::Gt
   };
 
-  const std::vector<BinExpr::Binary> EQ_OPS_BODY {BinExpr::Eq, BinExpr::Neq};
+  static const std::vector<BinExpr::Binary> EQ_OPS_BODY {BinExpr::Eq, BinExpr::Neq};
 
-  std::stringstream s;
+  for (size_t i = 0; i < COMP_OPS.size(); ++i) {
```

Comment:
Stupid question. Why can't we use some kind of an iterator over this vector here?

Reply:
I think we can. I just copy paste this piece of code. I will fix it.